### DISCUSSION TITLE
The robust optimization result is NaN

### DISCUSSION THREAD

**Discussion by Nash2Nash0** [yalmip/YALMIP#1244] [Problems using YALMIP]
A robust optimization model is built on matlab, and yalmip calls gurobi to solve it. The info shows that it was evaluated successfully and shows the optimal value in the process, but when viewed with the value function at the end, it shows NaN instead.

The robust optimization model is as follows:
p=sdpvar(T,1,'full');  %The scalar T was given.
x=sdpvar(size(C,2),1,'full');  %The matrix C was given.
F=[E*x<=f];   %The matrix E and the vector F were given.
W=[A*p<=b,uncertain(p)];   %The matrix A and the vector b were given, and the vector variable p was an uncertain variable.
obj=norm(C*x+d-p,1);
sol=optimize(F+W,obj)
obj=value(obj)
p=value(p)

The running results were shown in matlab as follows:
![image](https://user-images.githubusercontent.com/98147017/226509607-8b36c7d5-6e89-4efa-afd2-d7bacc61d0e5.png)

**Comment by johanlofberg** [comment]
A worst-case uncertainty is never computed (as there might not be a particular worst-case but a set, and the fact that it can be theoretically intractable to compute such a thing). It only computes a solution which is guaranteed to be feasible for all possible uncertainties

**Reply by johanlofberg** [reply]
and note the warning about 1-norms being conservatively handled (as it theoretically is intractable to derive the robust counterpart) and that you have to alter the option auxreduce to force YALMIP to derive the exact robust counterpart (whose complexity is exponential)

BTW, the use of 'full' is completely redundant here as you are defining vectors

**Reply by Nash2Nash0** [reply]
Thank you very much for your answer. Because the first time I used yalmip to solve a robust optimization problem. Although a worst-case uncertainty is never computed, the optimal objective function value does appear in the calculation process, but it still cannot be checked by the value function. May I ask how to deal with the problem that the optimal value and the optimal solution of the corresponding decision variable can be checked?

**Reply by johanlofberg** [reply]
To get the worst-case optimal value (assuming you've turned on auxreduce to actually compute the exact solution) you can introduce a new variable and then use `obj<=t `, and since it is robustly feasible it will hold the worst-case objective (this is what happens behind the scene)

**Reply by Nash2Nash0** [reply]
I just tried, and it's true that you can look at the worst-case by t, but it's actually more important for me to find the worst-case. Because my motivation for building robust optimization models is to find that worst-case scenario.Its mathematical model can be simply expressed as follows:
![image](https://user-images.githubusercontent.com/98147017/226541054-65e3ab6b-ea7a-47c4-ac0a-33b13ab98de9.png)
where, p is an uncertain variable, x is a decision variable, and other matrices and vectors are given.
How do you program in yalmip to get the worst case as accurately as possible?

**Reply by johanlofberg** [reply]
Now you switched the order, and this is much trickier as it would entail representing the minimum of a convex function which is very hard.

Don't you want to find the x which guarantees that the cost is as small as possible for the worst-case p?

**Reply by Nash2Nash0** [reply]
Yes, it is indeed very difficult. I converted its inner layer min into a single-layer mixed integer linear programming problem through duality theory and KKT before. However, due to the large scale of matrix E, solving the corresponding MILP problem becomes extremely time-consuming.
The reason I need to capture the worst case as much as possible is that the worst case will help me improve the scope identified in the first stage of my two-stage robust optimization model. In short, my focus now is to find the worst case of the robust optimization model expressed in the above mathematical form. However, I am very confused now. I don't know how to deal with this optimization problem. With your past experience, do you know how to deal with this optimization problem to get the worst case?

**Reply by johanlofberg** [reply]
Again, solving the problem of finding the x which minimizes the worst-case cost over p, is what YALMIP does for you, i.e. it solves the min_x max_p problem.

Given x, you can find a (not the) worst-case p by simply plugging in the optimal x, and then maximize the cost over p. This is a non-convex problem, but since the problem must be pretty small anyway (as you are able to solve the intractable worst-case certification) it should be easily solved. One can argue that it is a pretty silly approach though, as you basically do double work as one probably could do both in one shot by enumerating the uncertainty box and then analyze the solution of the enumerated model to figure out which vertices caused the epigraph of the objective to be active.

**Reply by Nash2Nash0** [reply]
Thank so much!
